### PR TITLE
[RFC] Add multi source input stream/demuxer

### PR DESCRIPTION
**Pull request by ace20022**
@FernetMenta 
Still a bit work to do, therefore RFC. Please ignore the [Util] changes.

**_Reading from demuxers:_**

The available demuxers are organized via a min heap. The sorting criteria is the dts or pts of the last packet they had delivered. The top demuxer in that heap is selected for reading.

**_Organizing the streams:_**

Unfortunately this does not work out of the box because each demuxer d_i numbers its streams from 0 to n_i, but the dvdplayer numbers them from 0 to sum_i(n_i). Hence the dvdplayer can't determine if the current packet should be forwarded to the audio/video player (because of the stream id mismatch). There are two solutions I can think of.  First we could copy the streams and collected them in the multi source demuxer. This is less future-proof and less efficient imo.I have implemented that as a poc here: https://github.com/xbmc/xbmc/compare/master...ace20022:ext3

The second one is presented by this pr. The streams of the demuxers are renumbered so that the ids match the ids of the dvdplayer. 
Before I put more work into it, which solution is preferable, or is there an even better one?

### CODE REVIEW

File: xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxFFmpeg.h [hunk 1]
```diff
@@ -146,7 +148,7 @@ class CDVDDemuxFFmpeg : public CDVDDemux
 
   CCriticalSection m_critSection;
   std::map<int, CDemuxStream*> m_streams;
-  std::vector<std::map<int, CDemuxStream*>::iterator> m_stream_index;
+  std::map<int, std::map<int, CDemuxStream*>::iterator> m_stream_index;
```
Comment (FernetMenta) [inline xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxFFmpeg.h]: a map is complex and a lookup is slow. for a small number of items a vector is much faster when being iterated and compared. I think we should avoid this double map and get rid of the other map too.

File: xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxMultiSource.cpp [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+}
+
+void CDVDDemuxMultiSource::GetStreamCodecName(int iStreamId, std::string &strName)
+{
+  auto iter = m_streamIdToDemuxerMap.find(iStreamId);
+  if (iter != m_streamIdToDemuxerMap.end())
+  {
+    iter->second->GetStreamCodecName(iStreamId, strName);
+  }
+};
```
Comment (FernetMenta) [inline xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxMultiSource.cpp]: ;

File: xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxMultiSource.cpp [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+int CDVDDemuxMultiSource::GetNrOfStreams()
+{
+  int streamsCount = 0;
+  for (auto iter : m_pDemuxers)
+    streamsCount += iter->GetNrOfStreams();
+
+  return streamsCount;
+}
+
+CDemuxStream* CDVDDemuxMultiSource::GetStream(int iStreamId)
```
Comment (FernetMenta) [inline xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxMultiSource.cpp]: those methods are called quite often. I am wondering if it makes sense to cache streams and update the cache on changes

File: xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxMultiSource.cpp [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+  return true;
+}
+
+void CDVDDemuxMultiSource::Reset()
+{
+  for (auto iter : m_pDemuxers)
+    iter->Reset();
+}
+
+DemuxPacket* CDVDDemuxMultiSource::Read()
```
Comment (FernetMenta) [inline xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxMultiSource.cpp]: that's the tricky part, isn't it? can you elaborate on how this works?

File: xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxMultiSource.cpp (hunk 3 above)
Comment (ace20022) [inline xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxMultiSource.cpp]: I have an alternative, but imo that's more error-prone. Cf.https://github.com/ace20022/xbmc/commit/4c0990f09e8e5e5ae12b0d28caef91b77d1db8aa

File: xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxMultiSource.cpp (hunk 4 above)
Comment (ace20022) [inline xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxMultiSource.cpp]: interestingly it's not, the managing of the streams, especially the ids, is the more problematic thing.
I'll add problem description to the pr post.

**Comment by ace20022** [comment]
@FernetMenta description added.

**Comment by FernetMenta** [comment]
nice work, well structured. The risk that it breaks existing behaviour is low so I think you can integrate this whenever you feel comfortable with it.
Regarding stream organisation, IMO the weak element in the chain is dvdplayer's class CSelectionStreams. The constraint of having no gaps and the very limited options of adding new streams from new sources makes life hard in the demuxers. If you consider to revamp this class go ahead :) up to you.

**Comment by ace20022** [comment]
> nice work, well structured. The risk that it breaks existing behaviour is low so I think you can integrate this whenever you feel comfortable with it.

Thanks ;)

> Regarding stream organisation, IMO the weak element in the chain is dvdplayer's class CSelectionStreams. The constraint of having no gaps and the very limited options of adding new streams from new sources makes life hard in the demuxers. If you consider to revamp this class go ahead :) up to you.

Maybe we can discuss this on irc at some point.

**Comment by ace20022** [comment]
@FernetMenta I was thinking about giving each stream a uuid. Imo that shouldn't be a problem performance wise, but needs a bigger rewrite. What do you think?

**Comment by FernetMenta** [comment]
are you talking about CSelectionStreams?

**Comment by ace20022** [comment]
yes, and as a consequence the demuxers' streams (or actually the other way round)

**Comment by FernetMenta** [comment]
CSelectionStreams really needs some rework and if you take this job, it is highly appreciated. IMO the biggest flaw of this class is that it does not allow gaps between streams. For this reason ffmpeg and pvr demuxers have become messy and error prone because on mpegts steams like an additional audio language may disappear. Currently those demuxers map a uid (PID) to a stream index, which is quite ugly.

Would be great if this could be simplified. I think this is what you mean by adding a uid, right?

**Comment by ace20022** [comment]
Yes, with uuid I mean using crossguid and the like

**Pull request closed without merging**